pkg/discovery/dtofactory/cluster_dto_builder.go

```diff
@@ -140,5 +150,59 @@ func (builder *clusterDTOBuilder) getCommoditiesSold(entityDTOs []*proto.EntityD
 		}
 		commoditiesSold = append(commoditiesSold, commSold)
 	}
-	return commoditiesSold, nil
+	return commoditiesSold, capacity, nil
+}
+
+// Create ContainerPlatformClusterData based on total namespace quota usage (which is equivalent to total container
+// resource capacity) and total node resource capacity.
+func (builder *clusterDTOBuilder) createClusterData(clusterName string, namespaceDTOs []*proto.EntityDTO,
+	nodeResourceCapacityMap map[proto.CommodityDTO_CommodityType]float64) *proto.EntityDTO_ContainerPlatformClusterData {
+	overcommitmentUsageMap := make(map[proto.CommodityDTO_CommodityType]float64)
+	for _, namespaceDTO := range namespaceDTOs {
+		for _, commodity := range namespaceDTO.GetCommoditiesSold() {
+			computeResourceCommType, commExists := quotaToComputeResourceCommMap[commodity.GetCommodityType()]
+			if commExists {
+				overcommitmentUsageMap[computeResourceCommType] += commodity.GetUsed()
+			}
+		}
+	}
+
+	vcpuOvercommitment := float64(0)
+	vmemOvercommitment := float64(0)
+	for commodityType, usage := range overcommitmentUsageMap {
+		switch commodityType {
+		case proto.CommodityDTO_VCPU:
+			resourceCapacity := builder.getNodeResourceCapacity(clusterName, commodityType, nodeResourceCapacityMap)
+			if resourceCapacity != 0 {
+				vcpuOvercommitment = usage / resourceCapacity
+			}
+		case proto.CommodityDTO_VMEM:
+			resourceCapacity := builder.getNodeResourceCapacity(clusterName, commodityType, nodeResourceCapacityMap)
+			if resourceCapacity != 0 {
+				vmemOvercommitment = usage / resourceCapacity
+			}
+		default:
+			glog.Errorf("Unsupported commodity type %s for cluster overcommitment", commodityType)
+		}
+	}
```

Comment:
Hmm... I'm wondering if it's important to be able to distinguish the case when there's no overcommitment (ie if the cluster is empty) vs if there's no usage in the quota. I'm thinking it might not be because actually getting 0 usage in practice would be extremely unusual. Just wanted to see if you had thought about this distinction. Like setting the overcommitment in the data to null if capacity == 0.

Maybe at least we could log a message if capacity is 0.

Reply:
Thanks @dblinn. This is a good point. Same 0 value could represent different meanings (empty cluster vs. limits not specified). As we discussed, we'll present overcommitment values together with other resource utilization data. We could identify if the cluster is empty based on resource utilization.

And I have added log messages if capacity is 0 in `getNodeResourceCapacity` func.